README.md
#  Overlay loader with app icon

A clean and light-weight package that shows a beautiful overlay loader with appIcon when initiating an async call.

## Demo App

![](https://media.giphy.com/media/9BIQDd4QuHEWihTuEz/giphy.gif)

## Getting started

[Example](https://github.com/esivuedaniel/OverlayLoaderWithAppIcon/blob/master/example/example.dart)

Add package to pubspec.yaml

```
dependencies:
  ...
  overlay_loader_with_app_icon:
```
Next, import the library into your widget.

```dart
import 'package:overlay_loader_with_app_icon/overlay_loader_with_app_icon.dart';
```

## Usage

To make an async call and display a modal with an appIcon progress indicator while the async call is taking place, wrap [OverlayLoaderWithAppIcon] around any widget.
Using the [isLoading] boolean variable, the progress loader can be enabled or disabled.

```dart

bool _loading = false;
@override
Widget build(BuildContext context) {
  return Scaffold(
    body: OverlayLoaderWithAppIcon(
        isLoading: _loading,
        appIcon:  Image.asset('images/appIcon.png'),
        child: Container(
       child: Form(...)
    )),
  );
}
```
## More Options
The current parameters are customizable in the constructor
```dart
OverlayLoaderWithAppIcon({
  required this.isLoading, 
  required this.child, 
  required this.appIcon, 
  this.appIconSize=50,
  this.borderRadius=15, 
  this.overlayOpacity=0.5, 
  this.circularProgressColor, 
  this.overlayBackgroundColor
});
```



## Thank you for using my package😊🙏🏻

## If you need more features
Please open an issue so that we can discuss it



CHANGELOG.md
## 0.0.1

* First release of Overlay_loader_with_app_icon, fully functional

## 0.0.2

* Update library versions, the sample and README.md

test/overlay_loader_with_app_icon_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:overlay_loader_with_app_icon/example.dart';

void main(){

  // testWidgets("load overlay when login button is clicked", (widgetTester)async{
  //   await widgetTester.pumpWidget(MaterialApp(home: LoginPage(),));
  //
  //   Finder findEmailTextField=find.byKey(ValueKey('email'));
  //   Finder findPasswordTextField=find.byKey(ValueKey('password'));
  //
  //   await widgetTester.enterText(findEmailTextField, "name@gmail.com");
  //   await widgetTester.enterText(findPasswordTextField, "Passw123@");
  //
  //   Finder loginButton=find.byType(ElevatedButton);
  //   await widgetTester.tap(loginButton);
  //   await widgetTester.pumpAndSettle();
  //
  //   Finder text_finder=find.text('Load');
  //
  //   expect(text_finder, findsNothing);
  //
  // });
}


example/example.dart

import 'package:flutter/material.dart';
import 'package:overlay_loader_with_app_icon/overlay_loader_with_app_icon.dart';



void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(primarySwatch: Colors.blue,),
      home:  LoginPage(),
    );
  }
}

class LoginPage extends StatefulWidget {
  const LoginPage({Key? key}) : super(key: key);

  @override
  State<LoginPage> createState() => _LoginPageState();
}

class _LoginPageState extends State<LoginPage> {

  // This maintains validators and state of form fields
  final GlobalKey<FormState> _formKey = GlobalKey<FormState>();

  TextEditingController emailController = TextEditingController();
  TextEditingController passwordController = TextEditingController();

  //This manage state of modal progress widget
  bool _isLoading=false;

  //Function to make an api call and wait for response or result
  void loadStateForSeconds(){
    setState(() {_isLoading=true;});  // start the modal progress HUD

    // Simulate a service call
    Future.delayed(Duration(seconds: 2), () {
      //Show result
      ScaffoldMessenger.of(context).showSnackBar( SnackBar(
        content:  Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          mainAxisSize: MainAxisSize.min,
          children: [
            Text('Login successfully', style: TextStyle(color: Colors.white, fontWeight: FontWeight.bold),),
            Text('${emailController.text}', style: TextStyle(color: Colors.white),),
          ],
        ),
        backgroundColor: Colors.blue,
      ));

      setState(() {_isLoading=false;}); // dismiss the modal progress HUD

    });
  }

  @override
  Widget build(BuildContext context) {
    //wrap widget with OverlayLoaderWithAppIcon
    return OverlayLoaderWithAppIcon(
      isLoading: _isLoading,
      overlayBackgroundColor: Colors.black,
      circularProgressColor: Color(0xff670099),
      appIcon:  Image.asset('images/appIcon.png'),
      child: Scaffold(
        appBar: AppBar(title: Text("Overlay loader with app icon"),),
        body: Form(
          key: _formKey,
          autovalidateMode: AutovalidateMode.always,
          child: Padding(
            padding: EdgeInsets.all(8.0),
            child: _buildColumn(),
          ),
        ),
      ),
    );
  }

  Column _buildColumn(){
    return Column(
      mainAxisAlignment: MainAxisAlignment.center,
      crossAxisAlignment: CrossAxisAlignment.center,
      children: <Widget>[
        TextFormField(
          controller: emailController,
          textInputAction: TextInputAction.next,
          decoration: InputDecoration(
              border: UnderlineInputBorder(),
              hintText: "Enter Email"
          ),
        ),

        SizedBox(height: 20,),
        TextFormField(
          controller: passwordController,
          obscureText: true,
          decoration: InputDecoration(
              border: UnderlineInputBorder(),
              hintText: "Enter Password"
          ),
        ),

        SizedBox(height: 20,),

        ElevatedButton(
            onPressed:(){
              if (_formKey.currentState!.validate()) {
                loadStateForSeconds();
              }
            },
            child: Text("Login"))
      ],
    );
  }

}



lib/overlay_animation.dart
import 'package:flutter/material.dart';

class OverLayAnimation extends StatefulWidget {

  final bool isLoading;
  final double opacity;
  final Color? color;
  final Widget progressIndicator;
  final Widget child;

  OverLayAnimation({
    required this.isLoading,
    required this.child,
    this.opacity = 0.5,
    this.progressIndicator = const CircularProgressIndicator(),
    this.color,
  });

  @override
  State<OverLayAnimation> createState() => _OverLayAnimationState();
}

class _OverLayAnimationState extends State<OverLayAnimation> with SingleTickerProviderStateMixin{
  late AnimationController _controller;
  late Animation<double> _animation;
  bool? _overlayVisible;

  _OverLayAnimationState();

  @override
  void initState() {
    super.initState();
    _overlayVisible = false;
    _controller = AnimationController(vsync: this, duration: Duration(milliseconds: 300));
    _animation = Tween(begin: 0.0, end: 1.0).animate(_controller);
    _animation.addStatusListener((status) {
      status == AnimationStatus.forward ? setState(() => {_overlayVisible = true}) : null;
      status == AnimationStatus.dismissed ? setState(() => {_overlayVisible = false}) : null;
    });
    if (widget.isLoading) {
      _controller.forward();
    }
  }

  @override
  void didUpdateWidget(OverLayAnimation oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (!oldWidget.isLoading && widget.isLoading) {
      _controller.forward();
    }

    if (oldWidget.isLoading && !widget.isLoading) {
      _controller.reverse();
    }
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    var widgets = <Widget>[];
    widgets.add(widget.child);

    if (_overlayVisible == true) {
      final modal = FadeTransition(
        opacity: _animation,
        child: Stack(
          children: <Widget>[
            Opacity(
              child: ModalBarrier(
                dismissible: false,
                color: widget.color ?? Theme.of(context).colorScheme.background,
              ),
              opacity: widget.opacity,
            ),
            Center(child: widget.progressIndicator),
          ],
        ),
      );
      widgets.add(modal);
    }

    return Stack(children: widgets);
  }
}


lib/overlay_loader_with_appicon.dart
library overlay_loader_with_app_icon;
import 'package:flutter/material.dart';


/// Inspired by https://pub.dev/packages/loading_overlay package
///
/// Wrapper [OverlayLoaderWithAppIcon] around any widget that makes an async call to show a modal with [appIcon] progress
/// indicator while the async call is in progress.
///
/// The progress indicator can be turned on or off using [isLoading]
///
/// The progress indicator defaults to a [CircularProgressIndicator] with [appIcon] inbetween,
/// the [appIcon] can be of any kind of widget
///
/// The appIcon size can be increased using [appIconSize]
/// The color of the modal barrier can be set using [overlayBackgroundColor]
///
/// The opacity of the modal barrier can be set using [overlayOpacity]
///
///The border radius of the container can be increased using[borderRadius]

class OverlayLoaderWithAppIcon extends StatelessWidget {
  final bool isLoading;
  final Widget child;
  final Widget appIcon;
  final double appIconSize;
  final double borderRadius;
  final double overlayOpacity;
  final Color? overlayBackgroundColor;
  final Color ?circularProgressColor;


  OverlayLoaderWithAppIcon({required this.isLoading, required this.child, required this.appIcon, this.appIconSize=50,
    this.borderRadius=15, this.overlayOpacity=0.5, this.circularProgressColor, this.overlayBackgroundColor});
  @override
  Widget build(BuildContext context) {
    return OverLayAnimation(
      isLoading: isLoading,
      opacity: overlayOpacity,
      color: overlayBackgroundColor??Theme.of(context).colorScheme.background,
      progressIndicator: Material(
        borderRadius: BorderRadius.circular(borderRadius),
        child: Padding(
          padding:  EdgeInsets.all(20),
          child: Stack(
            children: [
              Positioned(
                top: 0,
                bottom: 0,
                left: 0,
                right: 0,
                child: CircularProgressIndicator(
                  color:circularProgressColor??Theme.of(context).colorScheme.background,
                  strokeWidth: 2,
                ),
              ),
              SizedBox(
                width: appIconSize,
                height: appIconSize,
                child: appIcon,
              ),
            ],
          ),
        ),
      ),//Change this loading overlay
      child: child,
    );
  }
}


//OverLayAnimation class for ModalBarrier
class OverLayAnimation extends StatefulWidget {

  final bool isLoading;
  final double opacity;
  final Color? color;
  final Widget progressIndicator;
  final Widget child;

  OverLayAnimation({
    required this.isLoading,
    required this.child,
    this.opacity = 0.5,
    this.progressIndicator = const CircularProgressIndicator(),
    this.color,
  });

  @override
  State<OverLayAnimation> createState() => _OverLayAnimationState();
}

class _OverLayAnimationState extends State<OverLayAnimation> with SingleTickerProviderStateMixin{
  late AnimationController _controller;
  late Animation<double> _animation;
  bool? _overlayVisible;

  _OverLayAnimationState();

  @override
  void initState() {
    super.initState();
    _overlayVisible = false;
    _controller = AnimationController(vsync: this, duration: Duration(milliseconds: 300));
    _animation = Tween(begin: 0.0, end: 1.0).animate(_controller);
    _animation.addStatusListener((status) {
      status == AnimationStatus.forward ? setState(() => {_overlayVisible = true}) : null;
      status == AnimationStatus.dismissed ? setState(() => {_overlayVisible = false}) : null;
    });
    if (widget.isLoading) {
      _controller.forward();
    }
  }

  @override
  void didUpdateWidget(OverLayAnimation oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (!oldWidget.isLoading && widget.isLoading) {
      _controller.forward();
    }

    if (oldWidget.isLoading && !widget.isLoading) {
      _controller.reverse();
    }
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    var widgets = <Widget>[];
    widgets.add(widget.child);

    if (_overlayVisible == true) {
      final modal = FadeTransition(
        opacity: _animation,
        child: Stack(
          children: <Widget>[
            Opacity(
              child: ModalBarrier(
                dismissible: false,
                color: widget.color ?? Theme.of(context).colorScheme.background,
              ),
              opacity: widget.opacity,
            ),
            Center(child: widget.progressIndicator),
          ],
        ),
      );
      widgets.add(modal);
    }

    return Stack(children: widgets);
  }
}

lib/example/overlay_loader_with_app_icon.dart
library overlay_loader_with_app_icon;

import 'package:flutter/material.dart';
import 'package:overlay_loader_with_app_icon/overlay_loader_with_appicon.dart';



void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(primarySwatch: Colors.blue,),
      home:  LoginPage(),
    );
  }
}

class LoginPage extends StatefulWidget {
  const LoginPage({Key? key}) : super(key: key);

  @override
  State<LoginPage> createState() => _LoginPageState();
}

class _LoginPageState extends State<LoginPage> {

  // This maintains validators and state of form fields
  final GlobalKey<FormState> _formKey = GlobalKey<FormState>();

  TextEditingController emailController = TextEditingController();
  TextEditingController passwordController = TextEditingController();

  //This manage state of modal progress widget
  bool _isLoading=false;

  //Function to make an api call and wait for response or result
  void loadStateForSeconds(){
    setState(() {_isLoading=true;});  // start the modal progress HUD

    // Simulate a service call
    Future.delayed(Duration(seconds: 2), () {
      //Show result
      ScaffoldMessenger.of(context).showSnackBar( SnackBar(
        content:  Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          mainAxisSize: MainAxisSize.min,
          children: [
            Text('Login successfully', style: TextStyle(color: Colors.white, fontWeight: FontWeight.bold),),
            Text('${emailController.text}', style: TextStyle(color: Colors.white),),
          ],
        ),
        backgroundColor: Colors.blue,
      ));

      setState(() {_isLoading=false;}); // dismiss the modal progress HUD

    });
  }

  @override
  Widget build(BuildContext context) {
    //wrap widget with OverlayLoaderWithAppIcon
    return OverlayLoaderWithAppIcon(
      isLoading: _isLoading,
      overlayBackgroundColor: Colors.black,
      circularProgressColor: Color(0xff670099),
      appIcon:  Image.asset('images/appIcon.png'),
      child: Scaffold(
        appBar: AppBar(title: Text("Overlay loader with app icon"),),
        body: Form(
          key: _formKey,
          autovalidateMode: AutovalidateMode.always,
          child: Padding(
            padding: EdgeInsets.all(8.0),
            child: _buildColumn(),
          ),
        ),
      ),
    );
  }

  Column _buildColumn(){
    return Column(
      mainAxisAlignment: MainAxisAlignment.center,
      crossAxisAlignment: CrossAxisAlignment.center,
      children: <Widget>[
        TextFormField(
          key: ValueKey('email'),
          controller: emailController,
          textInputAction: TextInputAction.next,
          validator: (value) =>Validator.validateEmail(value!),
          decoration: InputDecoration(
              border: UnderlineInputBorder(),
              hintText: "Enter Email"
          ),
        ),

        SizedBox(height: 20,),
        TextFormField(
          key: ValueKey('password'),
          controller: passwordController,
          obscureText: true,
          validator: (value)=>Validator.validatePassword(value!),
          decoration: InputDecoration(
              border: UnderlineInputBorder(),
              hintText: "Enter Password"
          ),
        ),

        SizedBox(height: 20,),

        ElevatedButton(
            onPressed:(){
              if (_formKey.currentState!.validate()) {
                loadStateForSeconds();
              }
            },
            child: Text("Login"))
      ],
    );
  }

}

